### PR TITLE
Feat add patchflows debug functionality

### PR DESCRIPTION
**Pull request by prathamjagga**
I have implemented the debugging functionality as described in the roadmap. However, it has one issue that we are unable to take input from the user to continue execution. The problem might be occuring due to click or the rich terminal library being used. I'll appreciate if someone help me to fix it.

### CODE REVIEW

**Comment by codelion** [comment]
It will be hard to debug with the rich text library anyways, so when the debug flag is set you can disable the rich text formatting if that makes it easier.

**Comment by prathamjagga** [comment]
@codelion I've fixed the input part, But panel logs keep on displaying infinitely. I think this will be fixed once rich logging is disabled. Disabling rich logging will require changes at many places since rich text logging is the only way of logging at many places in the code. Do you want me to go ahead and make changes?

**Comment by codelion** [comment]
```
   File "/Users/codelion/Documents/GitHub/patchwork/patchwork/steps/ScanSemgrep/ScanSemgrep.py", line 31, in run                                       │
│     if self.inputs.get("debug") is not None:                                                                                                          │
│        ^^^^^^^^^^^                                                                                                                                    │
│ AttributeError: 'ScanSemgrep' object has no attribute 'inputs'                                                                                        │
│                                                                                                                                                       │
│ Error running patchflow AutoFix: 'ScanSemgrep' object has no attribute 'inputs' 
```

I am getting some error while trying our your branch.

**Comment by prathamjagga** [comment]
@codelion inputs were not initialized in some classes. Fixed now.

**Comment by prathamjagga** [comment]
I disabled the use of rich logging code wide. And debugger is working fine now. However wrote a method `init_debug_logging` in `logger.py` right after `init_cli_logging` . `init_debug_logging` is a copy of `init_cli_logging` for now. Please suggest what should be done in init_debug_logging.

**Comment by codelion** [comment]
Are all patchflows updated? I am getting this error:

```
 File "/Users/codelion/Documents/GitHub/patchwork/patchwork/patchflows/ResolveIssue/ResolveIssue.py", line 30, in __init__                                                
    if inputs.debug is None:                                                                                                                                               
       ^^^^^^^^^^^^                                                                                                                                                        
AttributeError: 'dict' object has no attribute 'debug'    
```

**Comment by codelion** [comment]
Also, I noticed that for some of the inputs can be really large like sarif_file or code2prompt. Printing them entirely everytime makes it impossible to follow the debug trace. May be we should truncate the inputs while printing so that they are not larger than a MAX_LENGTH(=1000).

**Comment by prathamjagga** [comment]
it was a syntax error, fixed. yeah lets truncate the inputs for now.

**Comment by prathamjagga** [comment]
Truncated the inputs while printing. Please review now. @codelion

**Comment by codelion** [comment]
can you please update from main to sync your fork as there are some conflicts now.

File: patchwork/steps/AnalyzeImpact/AnalyzeImpact.py [hunk 1]
```diff
@@ -84,6 +84,9 @@ def __init__(self, inputs: dict):
         self.inputs = inputs
 
     def run(self) -> dict:
+        if self.inputs.get("debug") is not None:
```
Comment (CTY-git) [inline patchwork/steps/AnalyzeImpact/AnalyzeImpact.py:87]: Instead of every implemented step having to check for this input key perhaps this should be managed in the abstract class `Step` at step.py instead.

File: patchwork/logger.py [hunk 2]
```diff
@@ -160,3 +160,28 @@ def init_cli_logger(log_level: str) -> logging.Logger:
     setattr(logger, "freeze", th.freeze)
     logger.setLevel(logging.TRACE)
     return logger
+
+def init_debug_logger(log_level: str) -> logging.Logger:
```
Comment (CTY-git) [inline patchwork/logger.py:164]: This function seems duplicated.

File: patchwork/app.py [hunk 3]
```diff
@@ -203,11 +209,62 @@ def cli(
...
+        if debug:
+            inputs["debug"] = True
+
         patchflow_class = find_patchflow(possbile_module_paths, patchflow_name)
         if patchflow_class is None:
             logger.error(f"Patchflow {patchflow_name} not found in {possbile_module_paths}")
             exit(1)
 
+    else:
+        with logger.panel("Initializing Patchwork CLI"):
```
Comment (CTY-git) [inline patchwork/app.py:221]: This seems duplicated as well whats the difference between the else block and the if block?

File: patchwork/app.py [hunk 4]
```diff
@@ -232,6 +291,22 @@ def cli(
             logger.debug(traceback.format_exc())
             logger.error(f"Error running patchflow {patchflow}: {e}")
             exit(1)
+    else:
+        with logger.panel(f"Patchflow {patchflow} logs") as _:
+            if inputs.get("debug") is not None:
```
Comment (CTY-git) [inline patchwork/app.py:296]: If the debug flag is not set there is no need to handle the debug value on this else block.

File: patchwork/app.py [hunk 5]
```diff
@@ -219,7 +276,9 @@ def cli(
             # treat --key=value as a key-value pair
             inputs[key] = value
 
-    with logger.panel(f"Patchflow {patchflow} logs") as _:
+    if debug: 
+        if inputs.get("debug") is not None:
```
Comment (CTY-git) [inline patchwork/app.py:280]: if the debug flag is set the debug value is always true if Im not wrong.

**Review by CTY-git — COMMENTED** [review]
Hi thanks for taking this up but I found multiple duplicate code which makes this hard to differentiate the difference between when the debug flag is set and when the debug flag is unset.

File: patchwork/logger.py (hunk 2 above)
Comment (prathamjagga) [reply patchwork/logger.py:164]: this function can be modified later on, if we want to make any changes in logging during debugging.

File: patchwork/app.py (hunk 3 above)
Comment (prathamjagga) [reply patchwork/app.py:221]: the difference is using logger.panel in case of debugging disabled and without logger.panel otherwise.

File: patchwork/app.py (hunk 5 above)
Comment (prathamjagga) [reply patchwork/app.py]: fixed.

**Comment by prathamjagga** [comment]
@CTY-git

File: patchwork/logger.py (hunk 2 above)
Comment (CTY-git) [reply patchwork/logger.py:164]: If we do need to change it later on, we can add it later.

File: patchwork/app.py (hunk 3 above)
Comment (CTY-git) [reply patchwork/app.py:221]: If that is the case I think it will be better if the `logger.panel` method can be assigned to a variable instead and is switched to a noop method when debug flag is true.

File: patchwork/steps/AnalyzeImpact/AnalyzeImpact.py (hunk 1 above)
Comment (CTY-git) [reply patchwork/steps/AnalyzeImpact/AnalyzeImpact.py:87]: Please do not let it be for now. Thank you.

**Review by CTY-git — CHANGES_REQUESTED** [review]

**Comment by prathamjagga** [comment]
@codelion I have made the requested changes. Please review now.

**Comment by codelion** [comment]
@prathamjagga if you can call the `debug` method in the base class then you won't need to edit every single step call it.

File: patchwork/steps/CreatePR/CreatePR.py [hunk 6]
```diff
@@ -18,6 +18,7 @@ class CreatePR(Step):
 
     def __init__(self, inputs: dict):
         super().__init__(inputs)
+        self.inputs = inputs
```
Comment (codelion) [inline patchwork/steps/CreatePR/CreatePR.py:21]: Why is this required?

File: patchwork/steps/CreatePR/CreatePR.py [hunk 7]
```diff
@@ -68,7 +69,7 @@ def __push(self, repo):
         logger.debug(f"Pushed to {self.original_remote_name}/{self.target_branch}")
         return is_push_success
 
-    def run(self) -> dict:
+    def run(self) -> dict:
```
Comment (codelion) [inline patchwork/steps/CreatePR/CreatePR.py:72]: Can you make sure there are no changes in whitespace like this.

**Comment by prathamjagga** [comment]
@codelion  you can try now.

**Comment by codelion** [comment]
@CTY-git this looks good now, is it fine to merge?

File: patchwork/app.py [hunk 8]
```diff
@@ -232,6 +242,21 @@ def cli(
             logger.debug(traceback.format_exc())
             logger.error(f"Error running patchflow {patchflow}: {e}")
             exit(1)
+    else:
+        with logger.panel(f"Patchflow {patchflow} logs") as _:
+            logger.info("DEBUGGING ENABLED. INPUTS WILL BE SHOWN BEFORE EACH STEP BEFORE PROCEEDING TO RUN IT.")
```
Comment (CTY-git) [inline patchwork/app.py:247]: debugging is false here, so this info logging should not be here.

File: patchwork/step.py [hunk 9]
```diff
@@ -90,11 +104,47 @@ def __managed_run(self, *args, **kwargs) -> Any:
...
+            old_settings = termios.tcgetattr(fd)
+            try:
+                tty.setraw(sys.stdin.fileno())
+                key = sys.stdin.read(1)
+            finally:
+                termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
+            return key
+        
+    def debug(self, inputs):
+        if inputs.get("debug") is None:
```
Comment (CTY-git) [inline patchwork/step.py:127]: This condition only checks for the existence of the `debug` input key. We should ensure that the value of the key `debug` is truthy as well.

File: patchwork/app.py [hunk 10]
```diff
@@ -232,6 +242,21 @@ def cli(
             logger.debug(traceback.format_exc())
             logger.error(f"Error running patchflow {patchflow}: {e}")
             exit(1)
+    else:
+        with logger.panel(f"Patchflow {patchflow} logs") as _:
```
Comment (CTY-git) [inline patchwork/app.py:246]: This is duplicated again, best to just reassign the `panel` var and reuse it here.

**Review by CTY-git — CHANGES_REQUESTED** [review]

**Comment by prathamjagga** [comment]
@CTY-git you can check now.

**Review by CTY-git — APPROVED** [review]
LGTM

**Comment by prathamjagga** [comment]
Feels good, getting your commits merged to such a beautiful project 🙂. 
Next time I'll take care of code quality and test everything well from the start only. Thankyou @codelion @CTY-git
